### PR TITLE
SWL-4344: New port features - AN and FEC Configurable & Conflict changes

### PR DESCRIPTION
**Pull request by harishraghavendra**
Reviewer: @poolakiran @kenchiang 

Implementation of new port features as per below design
https://bigswitch.atlassian.net/wiki/spaces/BSC/pages/889913608/Port+features+enhancement

### CODE REVIEW

**Comment by bsn-abat** [comment]
Find artifact changes from this pull request at https://github.com/floodlight/loxigen-artifacts/tree/PR-650-b1

File: openflow_input/bsn_port_desc_properties [hunk 1]
```diff
@@ -75,3 +75,17 @@ struct of_port_desc_prop_bsn_misc_capabilities : of_port_desc_prop_bsn {
...
+    uint32_t experimenter == 0x5c16c7;
+    uint32_t exp_type == 8;
+    uint8_t an_configurable;   /* AN Configurablity in HW */
+    uint8_t fec_configurable;  /* FEC Configurablity in HW */
+    uint8_t an_conflict;       /* AN Conflict with HW */
+    uint8_t fec_conflict;      /* FEC Conflict with HW */
+    uint64_t reserve1;         /* reserved for future */
+    uint64_t reserve2;         /* reserved for future */
+    uint64_t reserve3;         /* reserved for future */
+};
```
Comment (kenchiang) [inline openflow_input/bsn_port_desc_properties]: typo: "configurability"
do we need to use a uint8_t for each of the an_configurable, fec_configurable, an_conflict, fec_conflict fields?
maybe have a uint64_t named "configurability" and then define an enum similar to ofp_bsn_speed_capabilities with value 1 for AN, value 2 for FEC, value 4 for the next configurable property, etc.
also same idea of uint64_t named "conflict" which could use the same enum for "configurability".
suggest also renaming "reserve1" to "reserved1", etc.

**Comment by bsn-abat** [comment]
Find artifact changes from this pull request at https://github.com/floodlight/loxigen-artifacts/tree/PR-650-b2

File: openflow_input/bsn_port_desc_properties [hunk 2]
```diff
@@ -75,3 +75,28 @@ struct of_port_desc_prop_bsn_misc_capabilities : of_port_desc_prop_bsn {
...
+    OFP_BSN_EXT_CAP_CONFIGURABILITY_AN_DEFAULT = 0x1,
+    OFP_BSN_EXT_CAP_CONFIGURABILITY_AN_ENABLED = 0x2,
+    OFP_BSN_EXT_CAP_CONFIGURABILITY_AN_DISABLED = 0x4,
+    OFP_BSN_EXT_CAP_CONFIGURABILITY_FEC_DEFAULT = 0x8,
+    OFP_BSN_EXT_CAP_CONFIGURABILITY_FEC_ENABLED = 0x10,
+    OFP_BSN_EXT_CAP_CONFIGURABILITY_FEC_DISABLED = 0x20,
+};
+
+enum ofp_bsn_extended_capabilities_conflict(wire_type=uint64_t, bitmask=True) {
+    OFP_BSN_EXT_CAP_CONFLICT_AN = 0x1,
```
Comment (poolakiran) [inline openflow_input/bsn_port_desc_properties]: I think we can use one enum both for AN and FEC.

Comment (harishraghavendra) [reply openflow_input/bsn_port_desc_properties]: Addressed

File: openflow_input/bsn_port_desc_properties [hunk 3]
```diff
@@ -75,3 +75,28 @@ struct of_port_desc_prop_bsn_misc_capabilities : of_port_desc_prop_bsn {
     uint64_t supported;     /* All miscellaneous capabilities of port
                              * supported by HW */
 };
+
+enum ofp_bsn_extended_capabilities_configurability(wire_type=uint64_t, bitmask=True) {
+    OFP_BSN_EXT_CAP_CONFIGURABILITY_AN_DEFAULT = 0x1,
```
[the remark this answers is not in the document]
Comment (harishraghavendra) [reply openflow_input/bsn_port_desc_properties]: Addressed

**Comment by bsn-abat** [comment]
Find artifact changes from this pull request at https://github.com/floodlight/loxigen-artifacts/tree/PR-650-b3

File: openflow_input/bsn_port_desc_properties [hunk 4]
```diff
@@ -75,3 +75,19 @@ struct of_port_desc_prop_bsn_misc_capabilities : of_port_desc_prop_bsn {
...
+    OFP_BSN_EXT_CAP_FEC = 0x2,
+};
+
+struct of_port_desc_prop_bsn_extended_capabilities : of_port_desc_prop_bsn {
+    uint16_t type == 0xffff;
+    uint16_t length;
+    uint32_t experimenter == 0x5c16c7;
+    uint32_t exp_type == 8;
+    uint64_t configurability;   /* Bitmap of ofp_bsn_extended_capabilities_configurability in HW */
+    uint64_t conflict;          /* Bitmap of ofp_bsn_extended_capabilities_conflict with HW state*/
```
Comment (kenchiang) [inline openflow_input/bsn_port_desc_properties]: looks like the comments are now out-of-date...

**Comment by bsn-abat** [comment]
Find artifact changes from this pull request at https://github.com/floodlight/loxigen-artifacts/tree/PR-650-b4

**Comment by kenchiang** [comment]
looks good; if kiran doesn't have anything to add, please merge when ready

**Comment by poolakiran** [comment]
Do you see any downside in extending and using enum ofp_bsn_misc_capabilities instead of defining new enum for ext caps ? LGTM otherwise.

**Comment by harishraghavendra** [comment]
Thanks Ken & Kiran. 

I wanted to keep misc and ext capabilities as different enums as they both are defined in different structures.

**Pull request merged**